server/app/services/geo/esri/EsriClient.java

```diff
@@ -0,0 +1,179 @@
+  // response
+  public static final String ESRI_FIND_ADDRESS_CANDIDATES_OUT_FIELDS =
+      "Address, SubAddr, City, Region, Postal";
+  public static final String ESRI_FIND_ADDRESS_CANDIDATES_FORMAT = "json";
+  public Optional<String> ESRI_FIND_ADDRESS_CANDIDATES_URL;
+
+  @Inject
+  public EsriClient(Config configuration, WSClient ws) {
+    this.ws = ws;
+    this.ESRI_FIND_ADDRESS_CANDIDATES_URL =
+        Optional.ofNullable(configuration.getString("esri_find_address_candidates_url"));
+  }
+
+  /** Retries failed requests up to the provided value */
+  private CompletionStage<WSResponse> tryRequest(WSRequest request, MutableInt retries) {
+    CompletionStage<WSResponse> responsePromise = request.get();
+    responsePromise.handle(
+        (result, error) -> {
+          if (error != null || result.getStatus() != 200) {
+            if (retries.compareTo(new MutableInt(0)) > 0) {
+              retries.decrement();
+              return tryRequest(request, retries);
+            } else {
+              return responsePromise;
+            }
+          } else {
+            return CompletableFuture.completedFuture(result);
+          }
+        });
+
+    return responsePromise;
+  }
+
+  /**
+   * Returns address candidates from Esri's findAddressCandidates service
+   *
+   * <p>@see <a
+   * href="https://gisdata.seattle.gov/server/sdk/rest/index.html#/Find_Address_Candidates/02ss00000015000000/">Find
+   * Address Candidates</a>
+   */
+  public CompletionStage<Optional<JsonNode>> fetchAddressSuggestions(ObjectNode addressJson) {
+    if (this.ESRI_FIND_ADDRESS_CANDIDATES_URL.isEmpty()) {
+      return CompletableFuture.completedFuture(Optional.empty());
+    }
+    WSRequest request = ws.url(this.ESRI_FIND_ADDRESS_CANDIDATES_URL.get());
+    request.setContentType(ESRI_CONTENT_TYPE);
+    request.addQueryParameter("outFields", ESRI_FIND_ADDRESS_CANDIDATES_OUT_FIELDS);
+    // "f" stands for "format", options are json and pjson (PrettyJson)
+    request.addQueryParameter("f", ESRI_FIND_ADDRESS_CANDIDATES_FORMAT);
+    String address = addressJson.findPath(AddressField.STREET.getValue()).textValue();
+    System.out.println(address);
+    String address2 = addressJson.findPath(AddressField.LINE2.getValue()).textValue();
+    String city = addressJson.findPath(AddressField.CITY.getValue()).textValue();
+    String region = addressJson.findPath(AddressField.STATE.getValue()).textValue();
+    String postal = addressJson.findPath(AddressField.ZIP.getValue()).textValue();
+    if (address != null) {
+      request.addQueryParameter("address", address);
+    }
+    if (address2 != null) {
+      request.addQueryParameter("address2", address2);
+    }
+    if (city != null) {
+      request.addQueryParameter("city", city);
+    }
+    if (region != null) {
+      request.addQueryParameter("region", region);
+    }
+    if (postal != null) {
+      request.addQueryParameter("postal", postal);
+    }
+    MutableInt tries = new MutableInt(3);
```

Comment:
suggestion (non-blocking): consider making the number of tries configurable. If not configurable, extract to a constant at the top of the file for ease of reference

Reply:
Added a new configuration and updated EsriClient to read from that and default to 3 if null.